pkg/datapath/linux/devices.go

```diff
@@ -278,15 +288,139 @@ func (dm *DeviceManager) updateDevicesFromRoutes(l3DevOK bool, routes []netlink.
 
 		viable := dm.isViableDevice(l3DevOK, info.hasDefaultRoute, link)
 		if viable {
+			log.WithField(logfields.Device, name).Info("Added new device")
 			dm.devices[name] = struct{}{}
 			changed = true
+		} else {
+			log.WithField(logfields.Device, name).Debug("Skipping unviable device")
 		}
 	}
 	return changed
 }
 
+func (dm *DeviceManager) handleLinkUpdate(update netlink.LinkUpdate) bool {
+	if update.Header.Type != unix.RTM_DELLINK {
+		return false
+	}
+
+	name := update.Attrs().Name
+
+	if _, ok := dm.devices[name]; !ok {
+		return false
+	}
+
+	delete(dm.devices, name)
+
+	return true
+}
+
+// Listen starts listening to changes to network devices. When devices change the new set
+// of devices is sent on the returned channel.
+func (dm *DeviceManager) Listen(ctx context.Context) (chan []string, error) {
+	return dm.listen(ctx, nil)
+}
+
+// listen is the internal method to start listening for changes to network devices. This is split from the
+// public method to allow tests to listen to devices in specific network namespace.
+func (dm *DeviceManager) listen(ctx context.Context, netNS *netns.NsHandle) (chan []string, error) {
+	l3DevOK := supportL3Dev()
+	routeChan := make(chan netlink.RouteUpdate)
+	closeChan := make(chan struct{})
+
+	err := netlink.RouteSubscribeWithOptions(routeChan, closeChan,
+		netlink.RouteSubscribeOptions{
+			Namespace: netNS,
+			// List existing routes to make sure we did not miss changes that happened after Detect().
+			// FIXME(JM): Perhaps could consider combining Detect() and Listen() so we don't have to do this.
+			// Having this whole thing be fully event-driven is likely a better design as that forces consumers
+			// to work with it correctly.
+			ListExisting: true,
+		})
+	if err != nil {
+		return nil, err
+	}
+
+	linkChan := make(chan netlink.LinkUpdate)
+	err = netlink.LinkSubscribeWithOptions(linkChan, closeChan, netlink.LinkSubscribeOptions{
+		Namespace: netNS,
+		// FIXME(JM): What if a device is removed between Detect() and Listen()?
+		ListExisting: false,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	devicesChan := make(chan []string)
+
+	go func() {
+		// If a specific namespace is requested, lock the thread and set the
+		// threads namespace. This is currently only relevant for testing.
+		if netNS != nil {
+			runtime.LockOSThread()
+			defer runtime.UnlockOSThread()
+			if err := netns.Set(*netNS); err != nil {
+				log.WithError(err).Warnf("Failed to set network namespace")
+			}
+		}
+
+		// To avoid multiple reloads of the datapath when lots of routes are added in one go,
+		// collect the route updates into a batch and process the batch periodically.
+		//
+		ticker := time.NewTicker(dm.routeBatchingInterval)
```

Comment:
I'm not sure if this is the right place to mitigate this concern.

I presume that `updateDevicesFromRoutes()` should relatively quickly identify whether a route introduces a new device or not, thereby filtering down the events such that we shouldn't signal to the caller too frequently, only if there truly are new devices showing up.

If the _caller_ needs to react to many devices changing at once, and ensure that it doesn't trigger the datapath reload too frequently beacuse _that_ code is expensive, then that's the concern of the caller. Hence the caller should have the code to mitigate this concern, probably via a `pkg/trigger`. I'm actually mildly surprised that `TriggerReloadWithoutCompile()` directly reinitializes the datapath without a trigger/ratelimiter, I had previously figured it was supposed to be async & ratelimited.

This way, we can simplify this core code around handling the channels and the interaction between the different events that the goroutine listens for, we can avoid a constant tick, and the ratelimiting is tied directly to the logic that's expensive.

Reply:
(I note also that this ratelimiting is not extended to device deletes, so if there were many of those, then this code doesn't currently mitigate such a case).

Reply:
I wasn't really thinking about the downstream logic, but more about reducing the amount of work done here if a lot of routes are added when a device comes up. However now that I think about it, this feels prematurely optimized. I'll drop it.